test/cmd/runner/main.go

```diff
@@ -0,0 +1,133 @@
+/*
+Copyright 2021 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package main
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"os"
+	"os/exec"
+
+	_ "k8s.io/client-go/plugin/pkg/client/auth/gcp"
+)
+
+const resourcesEnvVar = "KNATIVE_RESOURCES_PATH"
+
+func main() {
+	os.Exit(run(os.Args[1:]))
+}
+
+func run(args []string) (exitCode int) {
+	var (
+		err       error
+		resources []*resourceInfo
+		state     *os.ProcessState
+		ctx       context.Context = setupSignals()
+	)
+
+	path := os.Getenv(resourcesEnvVar)
+	if path == "" {
+		path = "testdata/env"
+	}
+
+	if resources, err = loadFixtures(path); err != nil {
+		log("failed to load test resources %s", err)
+		exitCode = 1
+		return
+	}
+
+	// Defer so we can cleanup properly if a panic occurs
+	// when we run the test binary
+	defer func() {
+		// don't use the prior context since it could have been
+		// cancelled with an interrupt signal
+		err = revertResources(context.Background(), resources)
+		if err != nil {
+			exitCode = 1
+		}
+	}()
+
+	err = applyResources(ctx, resources)
```

Comment:
I like the pattern of using a `cancel` function to clean things up, so something like:
```go
  revert, err := applySUT()
  if err != nil { ... }
  defer revert()
```

Reply:
or even `t.Cleanup(revert)`

Reply:
there's no `t.Cleanup` in this context.

Subsequently something in `apply` could panic and that wouldn't allow to clean up the partial resources

Reply:
I made an upstream issue: https://github.com/golang/go/issues/44929